### PR TITLE
Add knob-ts adapter for inputs/outputs

### PR DESCRIPTION
**Pull request by expandrew**
This continues the work I did in #1 but makes it more generalized to support other input devices and outputs.

### CODE REVIEW

**Comment by expandrew** [comment]
There are still ongoing issues for how Raspbian assigns a driver to the PowerMate when connected (seems to assign `powermate` driver instead of one that works with `hidraw` which is preferred by `node-hid` to find the device from the script)

👉 Update README to include instructions for how to configure Raspbian to assign the correct driver to PowerMate with `udev` when it's connected

**Comment by expandrew** [comment]
I can't figure out how to make Raspbian assign a different driver to the PowerMate when it's plugged in, below are my stream of consciousness notes from all this nonsense:

> (Starting this session with a lot of tabs open from the previous day; I'm in the middle of a lot of thoughts and nothing has really worked yet so it would probably help to re-state what I know and what I've tried...)
> 
> ### Driver Binding
> https://lwn.net/Articles/143397/ Tried to use this tutorial but I was finding that I was unable to re-bind the device to another driver; was getting an error like `echo: write error: No such device`
> 
> ### Drivers baked into kernel
> I found where the drivers are specified for this vendor/product ID pair in the kernel code itself;  
> ```
>  $ cd /lib/modules/4.19.57-v7l+/
> $ cat modules.alias | grep 077D  
> ```
> I can't exactly do anything with this information because the kernel is already compiled so I can't change it without recompiling the kernel and that's a yak I don't want to shave
> 
> ### Blacklisting modules from loading to prevent automatic driver assignment
> I found out I can add an entry to modprobe blacklist like this: 
> ```
> echo "blacklist powermate" >> /etc/modprobe.d/raspi-blacklist.conf
> echo "install powermate /bin/false" >> /etc/modprobe.d/raspi-blacklist.conf
> ``` 
> to prevent the driver from being loaded when the kernel decides it needs to assign it to the device when it's plugged in
> 
> But I'm not sure what this helps if I can't assign the driver  I _do_ want since it has no driver assigned 
> 
> ### Use /dev/input path to control/use PowerMate
> Since the default driver does make a path at `/dev/input/event0` for the PowerMate, I wonder if there is a NodeJS library that I could use to read the inputs/signals from this? The issue with this approach is there wouldn't be a way for me to send LED signals to a "read-only" interface (which I presume this would be) 
> 
> ### Plugging in a USB mouse
> This does get me the `usbhid` driver in `lsusb -t` for that device, and it also adds a `/dev/hidraw0` entry and the "driver" is listed as hid-generic from the `tree /sys/bus/usb/devices/1-1.1.3:1.0` output
> 
> `///`
> 
> Next idea is to think about trying to dig into the `new_id` thing in `/sys/bus/usb/drivers/.../new_id` etc to maybe add the vendor/product ID for the PowerMate to the usbhid driver?
> 
> **This doesn't appear to work, I think maybe the PowerMate just really isn't a HID in the eyes of Linux**
> 
> macOS seems to find it fine, but there must really be some underlying driver differences here (you think? of course there are)
> 
> `///`
> 
> Maybe we move back to `libusb` now with the "deployed" version of knob-ts...
> 
> `///`
> 
> Yeah, this works, with the exception of the LED parts of the code I wrote that worked on macOS ... I don't know why this is
> 
> - [ ] ~~hidraw driver for PowerMate on Raspberry Pi~~ Fuck it, this isn't going to work; I need to just use `libusb` (this seems to still work)
> 
> `///`
> 
> Next I need to add some form of logging in order to actually see what's happening in there when the app is running, in order to fix the LED situation while using libusb, so I think that's next
> 
---


With all that, now I baked in the `libusb` driver part via HID ([see `node-hid`'s README here](https://github.com/node-hid/node-hid/#selecting-driver-type)) and added some logging via Node's standard `debug` package so now I can do `DEBUG=knob-ts:* node dist/index.js` when I run and I will see my debug outputs

So still on the list:
- [ ] Debug / fix the LED situation now that we are using `libusb`
- [ ] Figure out `pm2` to run the app continuously with more control
- [ ] Add a `nuimo` input
- [ ] Add better LED feedback on inputs for PowerMate (requires LED situation to be figured out though, hard to do this without knowing what's wrong)

**Comment by expandrew** [comment]
**Update: the LED issues were not LED issues but instead `node-sonos-ts` event listener and IP address auto-assignment issues**

**TL;DR: set this flag when running: `SONOS_LISTENER_HOST=10.0.1.10`**

Running the app with full logs (`DEBUG=*`) gives all the node-sonos-ts logs too - when I did this, I noticed this line and the `10.8.0.1` IP showing up a lot:
image

`10.8.0.x` are the IPs the Pi assigns to VPN clients via [openvpn](https://github.com/expandrew/media-cube/tree/master/openvpn) and that is `tun0` interface in `ifconfig` (instead of `wlan0` which I actually would want). The Sonos is on Wi-Fi, which is the `10.0.1.x` subnet (which has no access to the `10.8.0.x` addresses so it can't make a call that the listener will pick up)

I found this line in the `node-sonos-ts` code to allow me to override the default host via its assignment logic: https://github.com/svrooij/node-sonos-ts/blob/master/src/sonos-event-listener.ts#L73

So I will be running this with `SONOS_LISTENER_HOST=10.0.1.10` on the Pi from now on

**Comment by expandrew** [comment]
Added a basic `pm2` configuration but it feels a little overkill, and I'm not sure the installation/package.json situation is quite right (should this be a peerDependency? It should be installed globally but I don't necessarily want the app to install it when you run `npm install` because the Pi has limited disk space and I'm starting to run out)

**Comment by expandrew** [comment]
Cross-posted from https://github.com/expandrew/media-cube/pull/4#issue-546328181 
> > This is ugly but it mostly works... should probably refactor this and make it prettier `¯\_(ツ)_/¯ `
> 
> The main motivation here was that I was unable to start the script without the PowerMate plugged into my development machine, which was impeding my experience while writing up the Nuimo input class/implementation. So this is mostly just a chore, but it makes the library a little more robust so I'm fine with that.
> 
> This lets me still set up a `new PowerMate()` in the index without there necessarily being a PowerMate plugged in - the idea is I can define what a PowerMate _would_ do when it is plugged in, and it will become active once the USB device is available. Very narrow use case in production, very common situation in development for me. Scratching my own itch here.
> 
> Here's a video:
> 
> 
> https://user-images.githubusercontent.com/3157928/103259989-cdea2500-4969-11eb-8d4c-488ad28d9357.mp4

**Comment by expandrew** [comment]
Tonight's unload: 

Hoo, added a feature I needed for making the script start up without a PowerMate attached ... was bugging me in development while building the Nuimo input because the script wouldn't start without the PowerMate being found

Honestly, harder and more annoying than I expected... some learnings are here:
- `HID.devices()`/`new HID.HID()` are costly and slow, and they don't necessarily return what you expect if the OS hasn't finished its initialization process after plugging in the device ... I was originally trying to just do something like `usbDetect.on('add:1917:1040', setupHid())` but the `HID.devices()` call inside of `setupHid()` was returning empty / not showing the PowerMate, and I [was reading](https://github.com/node-hid/node-hid#cost-of-hiddevices-and-new-hidhid-for-detecting-device-plugunplug) about how polling `HID.devices()` (or otherwise relying on it to return in a timely manner) is probably not the best idea
- [Someone did attempt to write a `node-hid-async` library](https://github.com/hgmelectronics/node-hid-async) but it is only a subset of the real `node-hid` features - I thought about using it but it didn't have a `.read()` method that I was using and I didn't wanna deal with it
- I was kinda confused about how to use `async`/`await` within a TypeScript class; I came across [this article](http://ivanbatic.com/using-async-await-typescript-classes/) but it didn't really clear things up for me, and I don't think it would've worked the way I needed/wanted it to for my purposes - I was able to work around it, and I didn't have to modify the index API which was a goal of mine (I like it nice and clean to just say `new PowerMate()` plz and you get one whether it's plugged in or not)

---

Cross-posted from https://github.com/expandrew/media-cube/pull/4#issue-546328181 
> > This is ugly but it mostly works... should probably refactor this and make it prettier `¯\_(ツ)_/¯ `
> 
> The main motivation here was that I was unable to start the script without the PowerMate plugged into my development machine, which was impeding my experience while writing up the Nuimo input class/implementation. So this is mostly just a chore, but it makes the library a little more robust so I'm fine with that.
> 
> This lets me still set up a `new PowerMate()` in the index without there necessarily being a PowerMate plugged in - the idea is I can define what a PowerMate _would_ do when it is plugged in, and it will become active once the USB device is available. Very narrow use case in production, very common situation in development for me. Scratching my own itch here.
> 
> Here's a video:
> 
> 
> https://user-images.githubusercontent.com/3157928/103259989-cdea2500-4969-11eb-8d4c-488ad28d9357.mp4

**Comment by expandrew** [comment]
Tonight's unload: 

Hoo, added a feature I needed for making the script start up without a PowerMate attached ... was bugging me in development while building the Nuimo input because the script wouldn't start without the PowerMate being found

Honestly, harder and more annoying than I expected... some learnings are here:
- `HID.devices()`/`new HID.HID()` are costly and slow, and they don't necessarily return what you expect if the OS hasn't finished its initialization process after plugging in the device ... I was originally trying to just do something like `usbDetect.on('add:1917:1040', () => setupHid())` but the `HID.devices()` call inside of `setupHid()` was returning empty / not showing the PowerMate, and I [was reading](https://github.com/node-hid/node-hid#cost-of-hiddevices-and-new-hidhid-for-detecting-device-plugunplug) about how polling `HID.devices()` (or otherwise relying on it to return in a timely manner) is probably not the best idea
- [Someone did attempt to write a `node-hid-async` library](https://github.com/hgmelectronics/node-hid-async) but it is only a subset of the real `node-hid` features - I thought about using it but it didn't have a `.read()` method that I was using and I didn't wanna deal with it
- I was kinda confused about how to use `async`/`await` within a TypeScript class; I came across [this article](http://ivanbatic.com/using-async-await-typescript-classes/) but it didn't really clear things up for me, and I don't think it would've worked the way I needed/wanted it to for my purposes - I was able to work around it, and I didn't have to modify the index API which was a goal of mine (I like it nice and clean to just say `new PowerMate()` plz and you get one whether it's plugged in or not)

---

Cross-posted from https://github.com/expandrew/media-cube/pull/4#issue-546328181 
> > This is ugly but it mostly works... should probably refactor this and make it prettier `¯\_(ツ)_/¯ `
> 
> The main motivation here was that I was unable to start the script without the PowerMate plugged into my development machine, which was impeding my experience while writing up the Nuimo input class/implementation. So this is mostly just a chore, but it makes the library a little more robust so I'm fine with that.
> 
> This lets me still set up a `new PowerMate()` in the index without there necessarily being a PowerMate plugged in - the idea is I can define what a PowerMate _would_ do when it is plugged in, and it will become active once the USB device is available. Very narrow use case in production, very common situation in development for me. Scratching my own itch here.
> 
> Here's a video:
> 
> 
> https://user-images.githubusercontent.com/3157928/103259989-cdea2500-4969-11eb-8d4c-488ad28d9357.mp4

**Comment by expandrew** [comment]
The list now:

- [x] Allow me to start the app without PowerMate plugged in
- [ ] Add Nuimo input (parity with PowerMate's functions ... nothing extra yet? - using `knob-ts--nuimo` branch)
- [ ] Consider changing to double-tap/triple-tap for previous/next instead of press and turn; would allow me to put back group volume adjustment (push and turn to adjust all speakers, etc)
- [ ] LED feedback on PowerMate (pulse once or twice on specific inputs or something; definitely get rid of the indefinite pulse on grouped though, and fix whatever bug there is with the pulsing after play/pause when grouped)
- [ ] Raspberry Pi disk space is getting weird again and I should probably migrate to a bigger card ... figure this out

**Comment by expandrew** [comment]
I also have a [`blink(1)`](https://blink1.thingm.com/) on its way so what can I do with that? Maybe indicator for grouped? Or something else?

**Comment by expandrew** [comment]
Now an open PR #2 for the Nuimo inputs, I'm not convinced it's stable yet but this is my first attempt and I'll fix it in the morning when I have more space to test it

**Comment by expandrew** [comment]
Nuimo is in the mix now, the docs/demo stuff is on #5

**Comment by expandrew** [comment]
Re: adding `TRIPLE_PRESS` to PowerMate in d84322d6b52ce89a864483af4e4747a905ac9992

https://user-images.githubusercontent.com/3157928/103498575-47bf5a00-4e13-11eb-9849-45def06d26bb.mp4

**Comment by expandrew** [comment]
image

Added:

- [x] "wake up" feature to tell when the device is actually awake
- [x] glyphs for volume / group volume adjustment
- [x] better sensitivity for rotation

Things are a little slow sometimes ... there seems to be a delay on the display of the glyphs, presumably because it sends the glyphs from the Raspberry Pi rather than having them on the device somehow... not sure how I could get around this but probably just add back that debounce so I'm sending fewer inputs so there are fewer signals going back and forth? Will try it like it is for a bit and see if it bothers me

**Comment by expandrew** [comment]
Welp this has been in use for the past few weeks and I think it's stable enough to merge as-is

There are still minor annoyances, but I can fix those as I go
